src/Source/TMSSource.js

```diff
@@ -65,21 +65,33 @@ class TMSSource extends Source {
         this.isTMSSource = true;
 
         if (!source.extent) {
-            // default to the full 3857 extent
-            this.extent = new Extent('EPSG:3857',
-                -20037508.342789244, 20037508.342789244,
-                -20037508.342789255, 20037508.342789244);
+            // default to the global extent
+            this.extent = globalExtentTMS.get(source.projection);
         }
 
-        this.zoom = source.zoom || { min: 0, max: 18 };
+        this.zoom = source.zoom;
 
         this.isInverted = source.isInverted || false;
-
         this.format = this.format || 'image/png';
         this.url = source.url;
-        this.tileMatrixSet = source.tileMatrixSet;
-
         this.projection = CRS.formatToTms(source.projection);
+        this.tileMatrixSetLimits = source.tileMatrixSetLimits;
+
+        if (!this.zoom) {
+            if (this.tileMatrixSetLimits) {
+                const arrayLimits = Object.keys(this.tileMatrixSetLimits);
+                const size = arrayLimits.length;
+                const maxZoom = Number(arrayLimits[size - 1]);
+                const minZoom = maxZoom - size + 1;
+
+                this.zoom = {
+                    min: minZoom,
+                    max: maxZoom,
+                };
+            } else {
+                this.zoom = { min: 2, max: 20 };
```

Comment:
Don't forget to update the documentation above.

Reply:
changed